Sources/NIOCore/ByteBuffer-lengthPrefix.swift

```diff
@@ -0,0 +1,107 @@
+    ///     - writeMessage:
+    /// - Throws: If the number of bytes written during `writeMessage` can not be exactly represented as the given `Integer` i.e. if the number of bytes written is greater than `Integer.max`
+    /// - Returns: Number of total bytes written
+    @discardableResult
+    @inlinable
+    public mutating func writeLengthPrefixed<Integer>(
+        endianness: Endianness = .big,
+        as integer: Integer.Type,
+        writeMessage: (inout ByteBuffer) throws -> Int
+    ) throws -> Int  where Integer: FixedWidthInteger {
+        var totalBytesWritten = 0
+        
+        let lengthPrefixIndex = self.writerIndex
+        // Write a zero as a placeholder which will later be overwritten by the actual number of bytes written
+        totalBytesWritten += self.writeInteger(.zero, endianness: endianness, as: Integer.self)
+        
+        let messageLength = try writeMessage(&self)
+        totalBytesWritten += messageLength
+        
+        guard let lengthPrefix = Integer(exactly: messageLength) else {
+            throw LengthPrefixError.messageLengthDoesNotFitExactlyIntoRequiredIntegerFormat
+        }
+        
+        self.setInteger(lengthPrefix, at: lengthPrefixIndex, endianness: endianness, as: Integer.self)
+        
+        return totalBytesWritten
+    }
+}
+
+extension ByteBuffer {
+    /// Reads an `integer` from `self` and passes it to `readMessage`. 
+    /// It is checked that exactly the number of bytes specified in the length prefix is read during the call to `readMessage`.
+    /// - Parameters:
+    ///     - endianness: The endianness of the length prefix `Integer` in this `ByteBuffer` (defaults to big endian).
+    ///     - integer: the desired `Integer` type used to read the length prefix
+    ///     - readMessage: A closure that takes the length prefix, reads exactly that number of bytes from the given `ByteBuffer` and returns the result.
+    /// - Throws: if `readMessage` returns a non-nil value and does not read exactly the number of bytes passed to it.
+    /// - Returns: `nil` if the length prefix could not be read, 
+    ///            the length prefix is negative or
+    ///            the buffer does not contain enough bytes to read a message of this length.
+    ///            Otherwise the result of `readMessage`.
+    @inlinable
+    public mutating func readLengthPrefixed<Integer, Result>(
+        endianness: Endianness = .big,
+        as integer: Integer.Type,
+        readMessage: (Integer, inout ByteBuffer) throws -> Result?
+    ) throws -> Result? where Integer: FixedWidthInteger {
+        guard let messageLength = self.readInteger(endianness: endianness, as: Integer.self) else {
+            return nil
+        }
+        guard messageLength >= 0,
+              messageLength <= readableBytes else {
+            return nil
+        }
+        
+        let readStartIndex = self.readerIndex
+        guard let result = try readMessage(messageLength, &self) else {
```

Comment:
I think if we should `slice` ourselves here and pass the slice into the `readMessage` function. This will avoid the need to dance with indices, and will also let the various `readX` functions behave sensibly in the body of `readMessage`.

Reply:
+1 on passing in a slice. I think a separate `readLengthPrefixedSlice` would actually be a useful function to have (it could then be used as the base implementation for this function).

Reply:
Changed to use a `slice`. 
@glbrntt do you mean something like this:
```swift
mutating func readLengthPrefixedSlice<Integer>(
    endianness: Endianness = .big,
    as integer: Integer.Type
) -> ByteBuffer? where Integer: FixedWidthInteger {
    readLengthPrefixed(endianness: endianness, as: Integer.self) { _, slice in slice }
}
```

Reply:
More or less -- just the other way around:

```swift
public mutating func readLengthPrefixed<Integer, Result>(
    endianness: Endianness = .big,
    as integer: Integer.Type,
    readMessage: (Int, inout ByteBuffer) throws -> Result?
) rethrows -> Result? where Integer: FixedWidthInteger {
    var slice = self.readLengthPrefixedSlice(...)
    return try readMessage(slice.readableBytes, &slice)
}
```

Reply:
Added ✅